macleann/enchiridion-client
# EXPANDED SHOW SUPPORT!

# Major updates to allow searching for any show and adding any episode to users' playlists

Buckle up! Many, many changes here. Requires pulling the [server side's](https://github.com/macleann/enchiridion-server) most recent version of `main`.

### Views, views, views:
- Added valid routes for `/search`, including trailing url slugs for season and episode details
- Removed routes no longer needed (`/seasons` and trailing url slugs for season and episode details)

### Nav:
- Changed "Seasons" in navbar to "Search Shows", per the route changes I just mentioned, that now goes to `/search` instead

### New components:

within new `search` directory:
- `SearchBar.js`: Meat and potatoes of this PR, surprisingly barebones
- `SearchProvider.js`: Responsible for sending fetches to the server side for search-related queries
- `SearchDetail.js`: Functions as a detailed view of any show, imports and calls `Seasons` function of `Seasons.js` and passes the show's seasons to list all seasons

within `svgs` directory:
- `MagnifyingGlass.js`: exports <svg> that displays a magnifying glass for search bars across the site

### Minor changes:
- Added `SearchProvider` to `GodProvider.js`
- Added `resultId` variable via `useParams` to `EpisodeDetail.js` and also added to `getEpisodeByNumberFromTMDB` function parameters
- Added `resultId` argument to `getEpisodeByNumberFromTMDB` function definition and fetch url
- Added `resultId` variable via `useParams` to `SeasonDetail.js` and also added to `getSeasonBySeasonNumber` function parameters
- Changed where each episode is linked to in `SeasonDetail.js` to the singular number "episode" rather than "episodes" to make it more sensible when reading the url
- Added `resultId` argument to `getAllSeasons` and `getSeasonBySeasonNumber` function definitions and fetch urls in `SeasonProvider.js`

### Major Changes:

`Seasons.js`:
- This component is now being imported by `SearchDetail.js`, which is passing a list of seasons as `seasons` prop
- Removed the SeasonContext definitions
- Removed the isLoading, setIsLoading useState definitions
- Removed the useEffect that fetches the list of seasons
- Kept logic that displays if no seasons are found
- Changed where each season is linked to to the singular number "season" rather than "seasons" to make it more sensible when reading the url

`PlaylistForm.js`:
- Imported and defined relevant variables from `SearchContext`
- Imported `MagnifyingGlass` component for search bar
- Added `searchTerm`, `show`, `showId`, `displaySelectedShow`, `displaySelectedSeason` useState variables
- Changed `showEpisodeSelect` useState definition to `displayEpisodeSelect` for clarity
- Added a `searchTerm` useEffect that fetches search results (which are all tv shows) from the server
- Changed the useEffect that was fetching all seasons on page render to watch for changes in new `showId` variable and passes that id to `getAllSeasons` function, which now expects an id (see change in Minor Changes above)
- Changed the useEffect that was watching changes in `seasonNumber` to now also pass the previously set `showId` to `getSeasonBySeasonNumber` function. **AND** once that season is retrieved, modifying each episode in that season to have a `series_id` key set to the `showId`'s value as well as a `series_name` key set to the `show.name`
- Added a displayAirDate function that takes an object as a parameter. The function returns a variety of formatted dates, whether it's just the year or something more complete, which is dependent on the way the passed object's key is named.
- Added a search bar to the form, which prompts the `searchTerm` useEffect to run
- Added a select that only appears when the search bar is changed. Selecting a show from the dropdown list toggles the subsequent seasons select dropdown visibility as well as setting the `show` and `showId` variables in state
- Added a visibility to seasons select dropdown dependent on whether a show has been selected after searching
- Added `episode.series_name` to playlist's list of episodes that have been added so users can know what show the added episode is from in case they forget, etc.

### Styling changes:
- New custom `search-bar` tailwind class in `index.css`
- Changed width breakpoints and sizes on custom `card` tailwind class in `index.css`

### Bug fixes:

Fixes #30 Feature: Show list of seasons when a user clicks on a show in the search bar
Fixes #31 Feature: Add a TV show search bar to the "PlaylistForm.js" component

## Type of change


# How to Test?

Make sure you've followed all installation and setup instructions in the README. **Requires pulling the [server side's](https://github.com/macleann/enchiridion-server) most recent version of `main`.**

```
git fetch origin nm-searching-shows
git checkout nm-searching-shows
npm start
```

- [ ] Click `Search Shows` in the nav
- [ ] Search through shows
- [ ] Click on a show
- [ ] Click on a season
- [ ] Click on an episode title
- [ ] Log in
- [ ] Go to `Playlists` in the nav
- [ ] Click `Create Playlist`
- [ ] Search through shows
- [ ] Select a show in the dropdown
- [ ] Select a season
- [ ] Select an episode
- [ ] Repeat the above 4 steps as many times as you'd like
- [ ] Save the playlist
- [ ] Edit any playlist

# Checklist:
